### ISSUE TITLE
Cast_ptr_alignment false negative in trait method default impl

### ISSUE THREAD

**Issue by dtolnay** [rust-lang-nursery/rust-clippy#3440]
I believe Clippy should catch both of the following casts but it currently only flags the first one.

```rust
unsafe fn frob(bytes: *const u8) -> *const String {
    // Correctly caught by cast_ptr_alignment.
    bytes as *const String
}

trait Trait {
    unsafe fn frob(bytes: *const u8) -> *const Self where Self: Sized {
        // Should be caught but is not.
        bytes as *const Self
    }
}
```

Would have prevented this bug in Chrome OS: https://bugs.chromium.org/p/chromium/issues/detail?id=900962.

*`clippy 0.0.212 (f5d868c 2018-11-15)`*

**Comment by ada4a** [comment]
The alignment of `Self` is not known in general in a trait definition, and this lint doesn't currently fire when the type of one of the sides isn't known for sure. Additionally, the method is provided by the trait, so there won't even be a trait impl where there will be a concrete `Self` which we _would_ be able to lint.

But we could use precisely this argument -- that casting from `u8` to an arbitrary `Self` is problematic because `Self` has an unknown alignment -- to lint such provided trait methods.

@rustbot label S-needs-discussion
